planner/cascades/transformation_rules.go

```diff
@@ -1814,6 +1804,84 @@ func (r *EliminateOuterJoinBelowAggregation) isAggFuncAllDupAgnostic(agg *planne
 	return true
 }
 
+// EliminateOuterJoinBelowProjection eliminate the outer join which below projection.
+type EliminateOuterJoinBelowProjection struct {
+	baseRule
+}
+
+// NewRuleEliminateOuterJoinBelowProjection creates a new Transformation EliminateOuterJoinBelowProjection.
+// The pattern of this rule is `Projection->Join->X`.
+func NewRuleEliminateOuterJoinBelowProjection() Transformation {
+	rule := &EliminateOuterJoinBelowProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandJoin, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *EliminateOuterJoinBelowProjection) Match(expr *memo.ExprIter) bool {
+	joinType := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin).JoinType
+	return joinType == plannercore.LeftOuterJoin || joinType == plannercore.RightOuterJoin
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to eliminate outer join which below projection.
+func (r *EliminateOuterJoinBelowProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	proj := old.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	joinExpr := old.Children[0].GetExpr()
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
+
+	ok, innerChildIdx, outerGroup, innerGroup, outerUniqueIDs := prepareForEliminateOuterJoin(joinExpr)
+	if !ok {
+		return nil, false, false, nil
+	}
+
+	// only when proj only use the columns from outer table can eliminate outer join.
+	if !plannercore.IsColsAllFromOuterTable(proj.GetUsedCols(), outerUniqueIDs) {
+		return nil, false, false, nil
+	}
+
+	innerJoinKeys := join.ExtractJoinKeys(innerChildIdx)
+	contain, err := isInnerJoinKeysContainUniqueKey(innerGroup, innerJoinKeys)
```

Comment:
I think `isInnerJoinKeysContainUniqueKey` and `prepareForEliminateOuterJoin` will only be used in `EliminateOuterJoinBelowProjection` and `EliminateOuterJoinBelowAggregation`. How about move these two methods to a new struct like:
```golang
type outerJoinEliminator struct {
}

func (e *outerJoinEliminator) prepareForEliminateOuterJoin(...){}

type EliminateOuterJoinBelowProjection struct {
  baseRule
  outerJoinEliminator
}
```

Reply:
done